### PR TITLE
usnic: add support for handling flags to dgram send calls.

### PR DESCRIPTION
**Pull request by bturrubiates**
@goodell @jsquyres @prankurgupta 
Please review.

### CODE REVIEW

**Comment by jsquyres** [comment]
@bturrubiates Do you have any tests to know that these changes are correct?

**Comment by bturrubiates** [comment]
@jsquyres Not anything that I'm going to contribute, I've currently been testing by modifying `fi_dgram` and `fi_ud_pingpong`. I can work on trying to come up with something to contribute to fabtests.

File: prov/usnic/src/usdf.h [hunk 1]
```diff
@@ -295,6 +295,12 @@ struct usdf_ep {
 			int ep_sock;
 			struct usdf_av *ep_av;
 
+			/*
+			 * Move to ep_tx and ep_rx
```
Comment (goodell) [inline prov/usnic/src/usdf.h]: Please throw a `TODO` in that comment so that it's clearer.  Also, there's a lot of vertical space unnecessarily being used on the line above and below the text.  A single line `/* TODO: [...] */` comment would be better, IMO.

File: prov/usnic/src/usdf_dgram.c [hunk 2]
```diff
@@ -131,30 +131,32 @@ usdf_dgram_recvmsg(struct fid_ep *fep, const struct fi_msg *msg, uint64_t flags)
...
 	struct usdf_dest *dest;
 
 	ep = ep_ftou(fep);
-
 	dest = (struct usdf_dest *)(uintptr_t) dest_addr;
-	return _usdf_dgram_send(ep, dest, buf, len, context);
+
+	if (ep->ep_tx_dflt_signal_comp ||
+		ep->e.dg.tx_op_flags & FI_COMPLETION)
+		completion = USD_SF_SIGNAL;
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: You have no default value for `completion` and no `else` case here, so you will get undefined behavior when completions have not been requested.

File: prov/usnic/src/usdf_dgram.c [hunk 3]
```diff
@@ -167,40 +169,41 @@ usdf_dgram_senddata(struct fid_ep *fep, const void *buf, size_t len,
...
+
 	memcpy(hdr, &dest->ds_dest.ds_udp.u_hdr, sizeof(*hdr));
-	hdr->uh_udp.source =
-		qp->uq_attrs.uqa_local_addr.ul_addr.ul_udp.u_addr.sin_port;
 
-	len = sizeof(*hdr);
-	for (i = 0; i < count; i++) {
-		memcpy(copybuf + len, iov[i].iov_base, iov[i].iov_len);
+	for (size_t i = 0; i < count; i++) {
+		memcpy(copybuf + len, iov[i].iov_base,
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: Does this line really extend past 80 chars?  The line break is ugly if not, and the 2nd line is not indented enough if the line break is actually needed.

File: prov/usnic/src/usdf_dgram.c (hunk 2 above)
Comment (bturrubiates) [inline prov/usnic/src/usdf_dgram.c]: Good catch. This one is unique in that it's the only one that needs either an else clause or an initial value. Seem to have missed that.

File: prov/usnic/src/usdf_dgram.c (hunk 3 above)
Comment (bturrubiates) [inline prov/usnic/src/usdf_dgram.c]: Will fix, didn't notice.

File: prov/usnic/src/usdf_dgram.c [hunk 4]
```diff
@@ -209,112 +212,134 @@ _usdf_dgram_send_iov_copy(struct usdf_ep *ep, struct usd_dest *dest,
 	return 0;
 }
 
+static ssize_t
+_usdf_dgram_send_iov(struct usdf_ep *ep, const struct usd_dest *dest,
+	const struct iovec *iov, size_t count, void *context,
+	uint8_t cq_entry)
+{
+	struct iovec send_iov[USDF_DGRAM_MAX_SGE];
+
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: why the empty line here?

File: prov/usnic/src/usdf_dgram.c [hunk 5]
```diff
@@ -209,112 +212,134 @@ _usdf_dgram_send_iov_copy(struct usdf_ep *ep, struct usd_dest *dest,
...
- 	struct usd_qp_impl *qp;
-	struct usd_udp_hdr *hdr;
-	uint32_t last_post;
-	struct usd_wq_post_info *info;
-	uint8_t *copybuf;
-	size_t len;
-	struct iovec send_iov[USDF_DGRAM_MAX_SGE];
-	size_t i;
+	uint8_t completion;
+	size_t send_size = sizeof(struct usd_udp_hdr);
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: The style in the rest of the code is generally (though probably not 100% consistently) to only declare variables and then initialize them with later statements.  My preference is to continue that style unless there's a good reason to do otherwise.

File: prov/usnic/src/usdf_dgram.c [hunk 6]
```diff
@@ -209,112 +212,134 @@ _usdf_dgram_send_iov_copy(struct usdf_ep *ep, struct usd_dest *dest,
...
+		USDF_DBG_SYS(EP_DATA, "max inject length exceeded (%zu\n)",
+			len + sizeof(struct usd_udp_hdr));
+		return -FI_EINVAL;
+	}
+
+	/*
+	 * fi_inject never generates a completion
+	 */
+	return usd_post_send_one_copy(ep->e.dg.ep_qp, &dest->ds_dest, buf, len,
+		0, NULL);
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: indenting

File: prov/usnic/src/usdf_dgram.c [hunk 7]
```diff
@@ -209,112 +212,134 @@ _usdf_dgram_send_iov_copy(struct usdf_ep *ep, struct usd_dest *dest,
...
+			msg->iov_count, msg->context, completion);
+	} else if (flags & FI_INJECT) {
+		USDF_DBG_SYS(EP_DATA, "max inject length exceeded (%zu\n)",
+			send_size);
+		return -FI_EINVAL;
+	}
 
-	last_post = _usd_post_send_one(wq, hdr, len + sizeof(*hdr), 1);
+	return _usdf_dgram_send_iov(ep, dest, msg->msg_iov,
+		msg->iov_count, msg->context, completion);
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: indenting

File: prov/usnic/src/usdf_dgram.c [hunk 8]
```diff
@@ -468,62 +499,121 @@ usdf_dgram_prefix_send(struct fid_ep *fep, const void *buf, size_t len,
...
 {
-	return usdf_dgram_prefix_sendv(fep, msg->msg_iov, msg->desc, msg->iov_count,
-				(fi_addr_t)msg->addr, msg->context);
+	struct usdf_ep *ep;
+	struct usd_dest *dest;
+	uint8_t completion;
+	size_t send_size = sizeof(struct usd_udp_hdr);
+
+	ep = ep_ftou(fep);
+	dest = (struct usd_dest *)(uintptr_t)(fi_addr_t) msg->addr;
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: Is the `(fi_addr_t)` cast actually needed?  Isn't that already `msg->addr`'s type?

File: prov/usnic/src/usdf_dgram.c [hunk 9]
```diff
@@ -468,62 +499,121 @@ usdf_dgram_prefix_send(struct fid_ep *fep, const void *buf, size_t len,
...
+		return _usdf_dgram_send_iov_copy_prefix(ep, dest, msg->msg_iov,
+			msg->iov_count, msg->context, completion);
+	} else if (flags & FI_INJECT) {
+		USDF_DBG_SYS(EP_DATA, "max inject length exceeded (%zu\n)",
+			send_size);
+		return -FI_EINVAL;
+	}
+
+	return _usdf_dgram_send_iov_prefix(ep, dest, msg->msg_iov,
+		msg->iov_count, msg->context, completion);
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: indenting

File: prov/usnic/src/usdf_dgram.c [hunk 10]
```diff
@@ -468,62 +499,121 @@ usdf_dgram_prefix_send(struct fid_ep *fep, const void *buf, size_t len,
...
 	struct usd_dest *dest;
-	struct usd_wq *wq;
- 	struct usd_qp_impl *qp;
-	struct usd_udp_hdr *hdr;
-	uint32_t last_post;
-	struct usd_wq_post_info *info;
-	struct iovec send_iov[USDF_DGRAM_MAX_SGE];
-	size_t len;
-	unsigned i;
+
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: kill the extra blank line

File: prov/usnic/src/usdf_dgram.c [hunk 11]
```diff
@@ -468,62 +499,121 @@ usdf_dgram_prefix_send(struct fid_ep *fep, const void *buf, size_t len,
...
+	uint8_t completion;
+	size_t send_size = sizeof(struct usd_udp_hdr);
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: preference again here for separating decls from assignments

**Comment by bturrubiates** [comment]
Updated to reflect cosmetic changes. Working on an update for the array bounds thing that @goodell brought up. Give it one more go through, I haven't tested this updated version yet. I just know that it compiles, I'll be doing a bit of testing soon.

File: prov/usnic/src/usdf_dgram.c [hunk 12]
```diff
@@ -78,7 +78,7 @@ usdf_dgram_recv(struct fid_ep *fep, void *buf, size_t len,
 
 	index = qp->uq_rq.urq_post_index;
 	rxd.urd_context = context;
-	rxd.urd_iov[0].iov_base = (uint8_t *)ep->e.dg.ep_hdr_buf +
+	rxd.urd_iov[0].iov_base = (uint8_t *) ep->e.dg.ep_hdr_buf +
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: Needless whitespace change?

File: prov/usnic/src/usdf_dgram.c [hunk 13]
```diff
@@ -128,33 +128,35 @@ ssize_t
...
-		const void *buf, size_t len,  void *context)
-{
-	if (len <= USD_SEND_MAX_COPY - sizeof(struct usd_udp_hdr)) {
-		return usd_post_send_one_copy(ep->e.dg.ep_qp,
-			&dest->ds_dest, buf, len, USD_SF_SIGNAL, context);
-	} else {
-		return usd_post_send_one(ep->e.dg.ep_qp, &dest->ds_dest,
-			buf, len, USD_SF_SIGNAL, context);
-	}
+		msg->iov_count, (fi_addr_t) msg->addr, msg->context);
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: Another needless whitespace change?

File: prov/usnic/src/usdf_dgram.c [hunk 14]
```diff
@@ -128,33 +128,35 @@ ssize_t
...
 }
 
 ssize_t
 usdf_dgram_send(struct fid_ep *fep, const void *buf, size_t len, void *desc,
 		fi_addr_t dest_addr, void *context)
 {
-	struct usdf_ep *ep;
 	struct usdf_dest *dest;
+	struct usdf_ep *ep;
+	uint8_t completion = 0;
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: declaration/initialization separation

File: prov/usnic/src/usdf_dgram.c [hunk 15]
```diff
@@ -128,33 +128,35 @@ ssize_t
...
 	dest = (struct usdf_dest *)(uintptr_t) dest_addr;
-	return _usdf_dgram_send(ep, dest, buf, len, context);
+
+	if (ep->ep_tx_dflt_signal_comp ||
+		(ep->e.dg.tx_op_flags & FI_COMPLETION))
+		completion = USD_SF_SIGNAL;
+
+	if (len + sizeof(struct usd_udp_hdr) <= USD_SEND_MAX_COPY) {
+		return usd_post_send_one_copy(ep->e.dg.ep_qp, &dest->ds_dest,
+					      buf, len, completion, context);
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: Looks like spaces are used to line the indent up nicely.  While I agree this looks nicer, it's not the style we're trying to match here.  Tabs only for indents at the beginning of lines, please.  Just replace these 6 spaces with a hard tab.

File: prov/usnic/src/usdf_dgram.c [hunk 16]
```diff
@@ -167,40 +169,40 @@ usdf_dgram_senddata(struct fid_ep *fep, const void *buf, size_t len,
...
+
+	hdr = (struct usd_udp_hdr *) copybuf;
+
 	memcpy(hdr, &dest->ds_dest.ds_udp.u_hdr, sizeof(*hdr));
-	hdr->uh_udp.source =
-		qp->uq_attrs.uqa_local_addr.ul_addr.ul_udp.u_addr.sin_port;
 
-	len = sizeof(*hdr);
-	for (i = 0; i < count; i++) {
+	for (size_t i = 0; i < count; i++) {
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: Was there a good reason to convert the existing `unsigned i` declaration at the top of the file into this `size_t i` declaration in the `for` loop?

File: prov/usnic/src/usdf_dgram.c [hunk 17]
```diff
@@ -167,40 +169,40 @@ usdf_dgram_senddata(struct fid_ep *fep, const void *buf, size_t len,
...
+		const struct iovec *iov, size_t count, void *context,
+		uint8_t cq_entry)
 {
-	struct usd_wq *wq;
- 	struct usd_qp_impl *qp;
+	struct usd_wq_post_info *info;
 	struct usd_udp_hdr *hdr;
+	struct usd_qp_impl *qp;
+	struct usd_wq *wq;
+
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: extra blank line inserted for no reason

File: prov/usnic/src/usdf_dgram.c [hunk 18]
```diff
@@ -167,40 +169,40 @@ usdf_dgram_senddata(struct fid_ep *fep, const void *buf, size_t len,
...
 	uint32_t last_post;
-	struct usd_wq_post_info *info;
 	uint8_t *copybuf;
-	size_t len;
-	unsigned i;
+	size_t len = sizeof(struct usd_udp_hdr);
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: declaration/init separation

File: prov/usnic/src/usdf_dgram.c [hunk 19]
```diff
@@ -167,40 +169,40 @@ usdf_dgram_senddata(struct fid_ep *fep, const void *buf, size_t len,
...
 		len += iov[i].iov_len;
 	}
 
-	/* adjust lengths */
+	hdr->uh_udp.source =
+		qp->uq_attrs.uqa_local_addr.ul_addr.ul_udp.u_addr.sin_port;
 	hdr->uh_ip.tot_len = htons(len - sizeof(struct ether_header));
 	hdr->uh_udp.len = htons(len - sizeof(struct ether_header) -
-			sizeof(struct iphdr));
+		sizeof(struct iphdr));
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: indentation should be deeper, per our earlier discussion

File: prov/usnic/src/usdf_dgram.c [hunk 20]
```diff
@@ -209,115 +211,137 @@ _usdf_dgram_send_iov_copy(struct usdf_ep *ep, struct usd_dest *dest,
...
+	uint8_t cq_entry)
+{
+	struct iovec send_iov[USDF_DGRAM_MAX_SGE];
+	struct usd_wq_post_info *info;
+	struct usd_udp_hdr *hdr;
+	struct usd_qp_impl *qp;
+	struct usd_wq *wq;
+	uint32_t last_post;
+	uint8_t *copybuf;
+	size_t len = 0;
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: declaration/init separation

File: prov/usnic/src/usdf_dgram.c [hunk 21]
```diff
@@ -209,115 +211,137 @@ _usdf_dgram_send_iov_copy(struct usdf_ep *ep, struct usd_dest *dest,
...
+
+	qp = to_qpi(ep->e.dg.ep_qp);
+	wq = &qp->uq_wq;
+	copybuf = wq->uwq_copybuf + (wq->uwq_post_index * USD_SEND_MAX_COPY);
+	hdr = (struct usd_udp_hdr *) copybuf;
+	memcpy(hdr, &dest->ds_dest.ds_udp.u_hdr, sizeof(*hdr));
+
+	/* adjust lengths and insert source port */
+	hdr->uh_ip.tot_len = htons(len + sizeof(struct usd_udp_hdr) -
+				   sizeof(struct ether_header));
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: more mixed spaces/tabs

File: prov/usnic/src/usdf_dgram.c [hunk 22]
```diff
@@ -209,115 +211,137 @@ _usdf_dgram_send_iov_copy(struct usdf_ep *ep, struct usd_dest *dest,
...
+	send_size = sizeof(struct usd_udp_hdr);
 	dest = (struct usd_dest *)(uintptr_t) dest_addr;
-
-	len = 0;
-	for (i = 0; i < count; i++) {
-		len += iov[i].iov_len;
+	completion = ep->ep_tx_dflt_signal_comp ||
+		(ep->e.dg.tx_op_flags & FI_COMPLETION);
+
+	for (size_t i = 0; i < count; i++)
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: again, no need to move the `size_t i` declaration, it just adds churn to the patch

File: prov/usnic/src/usdf_dgram.c [hunk 23]
```diff
@@ -209,115 +211,137 @@ _usdf_dgram_send_iov_copy(struct usdf_ep *ep, struct usd_dest *dest,
...
+		return -FI_EINVAL;
+	}
+
+	return _usdf_dgram_send_iov(ep, dest, msg->msg_iov,
+				    msg->iov_count, msg->context, completion);
 }
 
 ssize_t
-usdf_dgram_inject(struct fid_ep *fep, const void *buf, size_t len,
+usdf_dgram_inject(struct fid_ep *fep, const void *buf, const size_t len,
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: this `const` is a bit silly, no?

File: prov/usnic/src/usdf_dgram.c [hunk 24]
```diff
@@ -376,7 +401,7 @@ usdf_dgram_prefix_recv(struct fid_ep *fep, void *buf, size_t len,
 
 	index = qp->uq_rq.urq_post_index;
 	rxd.urd_context = context;
-	rxd.urd_iov[0].iov_base = (uint8_t *)buf +
+	rxd.urd_iov[0].iov_base = (uint8_t *) buf +
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: pointless whitespace churn

File: prov/usnic/src/usdf_dgram.c [hunk 25]
```diff
@@ -403,7 +428,7 @@ usdf_dgram_prefix_recvv(struct fid_ep *fep, const struct iovec *iov,
 
 	rxd.urd_context = context;
 	memcpy(&rxd.urd_iov[0], iov, sizeof(*iov) * count);
-	rxd.urd_iov[0].iov_base = (uint8_t *)rxd.urd_iov[0].iov_base +
+	rxd.urd_iov[0].iov_base = (uint8_t *) rxd.urd_iov[0].iov_base +
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: ditto

File: prov/usnic/src/usdf_dgram.c [hunk 26]
```diff
@@ -209,115 +211,137 @@ _usdf_dgram_send_iov_copy(struct usdf_ep *ep, struct usd_dest *dest,
...
+	/*
+	 * fi_inject never generates a completion
+	 */
+	return usd_post_send_one_copy(ep->e.dg.ep_qp, &dest->ds_dest, buf, len,
+				      0, NULL);
 }
 
-ssize_t usdf_dgram_rx_size_left(struct fid_ep *fep)
+ssize_t
+usdf_dgram_rx_size_left(struct fid_ep *fep)
```
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: more unrelated formatting changes

**Comment by goodell** [comment]
@bturrubiates So as to avoid spamming everyone's mailbox more than I already have, I did not add a comment on every single line that I saw that was misformatted or involved an unnecessary change.  Please adjust all of the formatting per our earlier discussion and re-submit.  If the proper formatting is unclear, ping me offline and we can discuss.

**Comment by bturrubiates** [comment]
@goodell per our earlier discussion I was trying to make this more kernel style like. That's what those minor formatting changes are. They're all in their own commit as well since I knew you or @jsquyres would complain about that.

**Comment by bturrubiates** [comment]
@jsquyres @goodell 
Should be updated.

File: prov/usnic/src/usdf_dgram.c [hunk 27]
```diff
@@ -531,17 +621,72 @@ usdf_dgram_prefix_sendv(struct fid_ep *fep, const struct iovec *iov, void **desc
...
 		send_iov[0].iov_len -= USDF_HDR_BUF_ENTRY;
-		_usdf_dgram_send_iov_copy(ep, dest, send_iov, count, context);
+
+		return _usdf_dgram_send_iov_copy(ep, dest, send_iov, count,
+				context, ep->ep_tx_completion);
+	} else if (ep->e.dg.tx_op_flags & FI_INJECT) {
+		USDF_DBG_SYS(EP_DATA,
+				"given inject length (%zu) exceeds max inject length (%d)\n",
+				len, USD_SEND_MAX_COPY);
+		return -FI_ENOSPC;
```
Comment (bturrubiates) [inline prov/usnic/src/usdf_dgram.c]: I changed all of the `-FI_EINVAL` here to `-FI_ENOSPC` as that seemed more appropriate to me. 

@jsquyres @goodell Do you all agree?

File: prov/usnic/src/usdf_dgram.c [hunk 28]
```diff
@@ -531,17 +621,72 @@ usdf_dgram_prefix_sendv(struct fid_ep *fep, const struct iovec *iov, void **desc
...
+				msg->iov_count);
+		return -FI_ENOSPC;
+	}
+
+	if (len <= USD_SEND_MAX_COPY) {
+		/* _usdf_dgram_send_iov_copy isn't prefix aware and allocates
+		 * its own prefix. reorganize iov[0] base to point to data and
+		 * len to reflect data length.
+		 */
+		memcpy(send_iov, msg->msg_iov,
```
Comment (bturrubiates) [inline prov/usnic/src/usdf_dgram.c]: Is this memcpy necessary? I don't think I should be modifying the input, but I could avoid doing a copy if so.

File: prov/usnic/src/usdf_dgram.c (hunk 27 above)
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: I don't think it matters too much.  I usually think of `ENOSPC` as indicating a lack of _disk_ space, but I'm sure I've probably used it to have your meaning in code that I've written in the past.  `ENOBUFS` is another choice here, but it's all bike-shedding, so we should probably just leave it as `ENOSPC` in your PR.

File: prov/usnic/src/usdf_dgram.c (hunk 28 above)
Comment (goodell) [inline prov/usnic/src/usdf_dgram.c]: Yes, it's needed because the `msg` is `const`.

Comment (bturrubiates) [inline prov/usnic/src/usdf_dgram.c]: That's what I figured...

**Comment by goodell** [comment]
This generally looks good, we just need to come to a resolution [on this issue](https://github.com/bturrubiates/libfabric/commit/940db1859e6c7e451a4bf9e8bce2654b58df7540#commitcomment-12088251) and run some performance checks to make sure things haven't slowed down unexpectedly from the refactoring/additions.

File: prov/usnic/src/usdf_dgram.c (hunk 27 above)
Comment (bturrubiates) [inline prov/usnic/src/usdf_dgram.c]: Alright sounds good.

File: prov/usnic/src/usdf_dgram.c [hunk 29]
```diff
@@ -124,204 +282,146 @@ usdf_dgram_recvv(struct fid_ep *fep, const struct iovec *iov, void **desc,
...
-
-	last_post = _usd_post_send_one(wq, hdr, len, 1);
-
-	info = &wq->uwq_post_info[last_post];
-	info->wp_context = context;
-	info->wp_len = len;
+	/* _usdf_dgram_send_iov allocates buffer of size USDF_DGRAM_MAX_SGE
+	 * but first slot is for header.
+	 */
+	if (count > (USDF_DGRAM_MAX_SGE - 1)) {
```
Comment (bturrubiates) [inline prov/usnic/src/usdf_dgram.c]: From bturrubiates/libfabric@5fde956c commit diff:

>Also, should this comparison be against USDF_DGRAM_DFLT_SGE (or some value derived from it) or txattr->iov_limit?

>In general though, fixing that needn't be a blocker for accepting this change. This is still a big improvement over where we were without this PR.

How about I move that check into `_usdf_dgram_send_iov` and change it to use `txattr->iov_limit` then change the way `iov_limit` is being set in a separate PR. Currently, it's being set to `USDF_DGRAM_DFLT_SGE` if no hint is provided.

I think it should be `USDF_DGRAM_DFLT_SGE` for prefix mode and `USDF_DGRAM_DFLT_SGE - 1` for non-prefix mode. Thoughts?

**Comment by bturrubiates** [comment]
@jsquyres @goodell 
another attempt. there should be one more commit coming, but please review again.

**Comment by bturrubiates** [comment]
Doh! I made a mistake....
Going to leave a note here for a fix I need to do: 

Take small message optimization out of prefix_send and put it in non-prefix send.

**Comment by goodell** [comment]
I still see several locations where `ep->ep_tx_completion` is passed directly to a `usd_post_send_XXX` routine without the ternary to convert to `USD_SF_SIGNAL` as needed.

**Comment by bturrubiates** [comment]
> I still see several locations where ep->ep_tx_completion is passed directly to a usd_post_send_XXX routine without the ternary to convert to USD_SF_SIGNAL as needed.

I only found one instance, and I've fixed it.

**Comment by bturrubiates** [comment]
Should have addressed your most recent comments.

**Comment by goodell** [comment]
:+1:
